find_exponential/src/main.rs

```diff
@@ -0,0 +1,197 @@
+/// If parsing takes longer than this many milliseconds, it is assumed to be non-linear.
+const MAX_MILLIS: u128 = 500;
+/// Size of repetition in bytes to parse.
+const NUM_BYTES: usize = 256*1024;
+/// Number of samples per pattern tested.
+const SAMPLE_SIZE: usize = 5;
+/// Score function to use when figuring out if something is non-linear.
+/// `scoring::{slope_stddev,pearson_correlation}`
+const SCORE_FUNCTION: fn(&Array2<f64>) -> (f64, bool) = scoring::slope_stddev;
+/// If slope_stddev is used, if the standard deviation is larger than this, it's assumed to be non-linear.
+const ACCEPTANCE_STDDEV: f64 = 20.0;
+/// If pearson_correlation is used, if the correlation coefficient is below this value,
+/// it's assumed to be non-linear.
+const ACCEPTANCE_CORRELATION: f64 = 0.995;
+/// 0 / 1 / 2 / 3
+const DEBUG_LEVEL: u8 = 2;
+
+fn main() {
+    let num_cpus = num_cpus::get() * 3 / 4;
+
+    let arg = env::args().nth(1);
+    if let Some("--retest") = arg.as_ref().map(|s| s.as_str()) {
+        for pattern in io::stdin().lock().lines() {
+            let pattern = pattern.unwrap();
+            println!("retesting {:?}", pattern);
+            test(&pattern);
+        }
+        return;
+    } else {
+        fuzz(num_cpus);
+    }
+}
+
+fn fuzz(num_cpus: usize) {
+    // get all literals from pulldown-cmark's source code
+    let walkdir = WalkDir::new("../src")
+        .into_iter()
+        .filter_map(|e| e.ok())
+        .filter(|e| e.file_type().is_file())
+        .filter(|e| if let Some(ext) = e.path().extension() { ext == "rs" } else { false });
+
+    let mut literals = HashSet::new();
+    for file in walkdir {
+        if file.path() == Path::new("../src/entities.rs") {
+            continue;
+        }
+        println!("parsing {}", file.path().display());
+        let content = fs::read_to_string(file.path())
+            .expect(&format!("unable to read file {:?}", file.path()));
+        let parsed = syn::parse_file(&content)
+            .expect(&format!("unable to parse file {:?}", file.path()));
+        literals::extract_literals_from_items(&mut literals, parsed.items);
+    }
+
+    let literals: Vec<_> = literals.into_iter()
+        .filter(|lit| !lit.contains(&b'\n'))
+        .filter(|lit| !lit.is_empty())
+        .collect();
+
+    let pool = ThreadPool::new(num_cpus);
+    let mut count = -(num_cpus as isize) * BATCH_SIZE as isize;
+    let start_time = Instant::now();
+    let (tx, rx) = mpsc::channel();
+    for _ in 0..num_cpus {
+        tx.send(()).unwrap();
+    }
+    loop {
+        rx.recv().unwrap();
+        let mut literals = literals.clone();
+        let tx = tx.clone();
+        count += BATCH_SIZE as isize;
+        pool.execute(move || {
+            literals.shuffle(&mut rand::thread_rng());
+            let combs = literals.iter()
+                .combinations(COMBINATIONS)
+                .take(BATCH_SIZE);
+            for comb in combs {
+                let res = panic::catch_unwind(|| {
```

Comment:
I don't know how exactly the `threadpool` handled panicking in threads. `rayon` afaict removes the thread from the thread pool, reducing the number of concurrent threads. To prevent that behaviour, and also not depend on any implementation detail of the implementation of the used thread pool, I'm catching the panic.
The `process::exit` is probably wrong though and instead it should just be logged.

Reply:
This was my oversight. I thought you were catching panics caused by potential errors in forming utf-8 strings.

Catching panics is essential for reporting bugs in pulldown-cmark :+1: